add: Performance Options & automatic optimization

morganholly opened the pull request:
add ability to run every N frames
add ability to change frame divisor based on forces applied (makes it mostly invisible)

only tested with small graph, but this resulted in going from maxing out my potato gpu and thermal throttling my cpu to having no visible hit

morganholly committed:
add some performance settings

add ability to run every N frames
add ability to change frame divisor based on forces applied (makes it mostly invisible)

only tested with small graph, but this resulted in going from maxing out my potato gpu and thermal throttling my cpu to having no visible hit

morganholly commented:
the automatic option to run less often when the graph is not changing much ends up running at only once every 20 frames on the small graph after it stabilizes. on larger graphs the calculation might break down, but just changing the 8/total_force to a different number would change that and it would be a pretty simple addition to add another export parameter

ThePat02 commented on addons/force_directed_graph/fdg_node.gd:
Add and explain the return value to the docstring.

ThePat02 commented on addons/force_directed_graph/fdg_spring.gd:
Same here! Just for keeping consistency!

ThePat02 commented on addons/force_directed_graph/force_directed_graph.gd:
```diff
@@ -15,7 +15,12 @@ var springs: Array[FDGSpring]
+## Run every N frames
+@export_range(0, 20) var frame_divide: int = 1
+## Use total applied force to calculate how often to update
+@export var auto_frame_divide: bool = true
```
Mention that this overwrites the value of `frame_divide`!

ThePat02 commented on addons/force_directed_graph/force_directed_graph.gd:
```diff
@@ -15,7 +15,12 @@ var springs: Array[FDGSpring]
+## Use total applied force to calculate how often to update
+@export var auto_frame_divide: bool = true
 
+var frame_counter = 1
```
I think a name more verbose like "frames_since_calculation" or similar would improve readability.

ThePat02 commented on project.godot:
I will update to a newer Godot version in another commit. Doesn't make any difference tho.

ThePat02 reviewed (changes requested):
Thank your for this PR! Besides some _minor_ naming and documentation changes, this is a great addition for the plugin. If you find some time to adjust your PR accordingly, I will gladly merge it into main!

Always happy to see people still use my old plugins :) (Especially **without** frying their GPU!!!)

morganholly commented:
ah yea that makes sense, ill work on that today

morganholly committed:
fixes for review changes requested

morganholly committed:
add compensation for lowered frame rate

morganholly committed:
check for additional movement (editor drag) to update faster

morganholly commented:
i added that additional documentation, i made it scale up forces when it runs less often (tho nonlinearly, to reduce jitter at very large divisions), and did some checking to look for editor dragging.

editor drag checking was the hardest part, since there was no notification or signal that would work as needed, but the checks i added do improve it a lot. it only checks that every 10 frames, which makes it not as smooth as it could be, but also reduces the performance hit. and when the drag gets going, it should update the total applied force enough to get it running most if not all frames.

morganholly commented:
also i updated the range on the frame divide to go up to 500, which it is also now capped at in the code. it generally seems to stabilize between 100 and 500 when in a stable state.

ThePat02 committed:
Update project.godot

ThePat02 commented:
@morganholly Thank you for your contribution <3 Great work :)

morganholly commented:
i will have to check it later but i just realized it might have a bug when the auto divide option is off, some of it at least still runs, i think
